zerver/views/auth.py

```diff
@@ -933,6 +919,96 @@ def api_fetch_api_key(
     RequestNotes.get_notes(request).requestor_for_logs = user_profile.format_requestor_for_logs()
 
     api_key = get_api_key(user_profile)
+    return api_key
+
+
+@csrf_exempt
+@has_request_variables
+def jwt_fetch_api_key(
+    request: HttpRequest,
+    json_web_token: str = REQ(default=""),
+    include_profile: bool = REQ(default=False, json_validator=check_bool),
+) -> HttpResponse:
+    return_data: Dict[str, bool] = {}
+
+    realm = get_realm_from_request(request)
+    if realm is None:
+        raise InvalidSubdomainError()
+
+    try:
+        key = settings.JWT_AUTH_KEYS[realm.subdomain]["key"]
+        algorithms = settings.JWT_AUTH_KEYS[realm.subdomain]["algorithms"]
+    except KeyError:
+        raise JsonableError(_("Auth key for this subdomain not found"))
+
+    try:
+        json_web_token = request.POST["json_web_token"]
+        options = {"verify_signature": True}
+        payload = jwt.decode(json_web_token, key, algorithms=algorithms, options=options)
+    except KeyError:
+        raise JsonableError(_("No JSON web token passed in request"))
+    except jwt.InvalidTokenError:
+        raise JsonableError(_("Bad JSON web token"))
+
+    remote_email = payload.get("email", None)
+    if remote_email is None or not remote_email:
+        raise JsonableError(_("No email specified in JSON web token claims"))
```

Comment:
Should we run an email address validator here? I think the answer is no, since we only do that in account creation normally, and this is not an account creation code path.

I'm also not sure we need the `or not remote_email` code here; `""` seems equally broken as an email address as say, `" "`, so unless there's a reason we'd expect that to be a common failure mode, I feel like the `Invalid email` exception that `authenticate` is going to be just as useful?

Reply:
> Should we run an email address validator here? I think the answer is no, since we only do that in account creation normally, and this is not an account creation code path.

Yeah that was my thinking here too

> I'm also not sure we need the `or not remote_email` code here; `""` seems equally broken as an email address as say, `" "`, so unless there's a reason we'd expect that to be a common failure mode, I feel like the `Invalid email` exception that `authenticate` is going to be just as useful? 

I'm guessing that the idea was that the empty string is failing to submit a value as opposed to something malformed like `" "` - but the `Invalid email` treatment makes sense to me too. Perhaps  @atoppi can you comment your idea regarding this?

Reply:
The `or not remote_email` part can be removed, it was just an extra check that we do not need anymore.

Reply:
Cool, pushed a commit cleaning that up.